examples/decoupled/square_model.py

```diff
@@ -0,0 +1,189 @@
+            model_config)
+        if not using_decoupled:
+            raise pb_utils.TritonModelException(
+                """the model `{}` can generate any number of responses per request,
+                enable decoupled transaction policy in model configuration to 
+                serve this model""".format(args['model_name']))
+
+        # Get IN configuration
+        in_config = pb_utils.get_input_config_by_name(model_config, "IN")
+
+        # Validate the shape and data type of IN
+        in_shape = in_config['dims']
+        if (len(in_shape) != 1) or (in_shape[0] != 1):
+            raise pb_utils.TritonModelException(
+                """the model `{}` requires the shape of 'IN' to be
+                [1], got {}""".format(args['model_name'], in_shape))
+        if in_config['data_type'] != 'TYPE_INT32':
+            raise pb_utils.TritonModelException(
+                """the model `{}` requires the data_type of 'IN' to be
+                'TYPE_INT32', got {}""".format(args['model_name'],
+                                               in_config['data_type']))
+
+        # Get OUT configuration
+        out_config = pb_utils.get_output_config_by_name(model_config, "OUT")
+
+        # Validate the shape and data type of OUT
+        out_shape = out_config['dims']
+        if (len(out_shape) != 1) or (out_shape[0] != 1):
+            raise pb_utils.TritonModelException(
+                """the model `{}` requires the shape of 'OUT' to be
+                [1], got {}""".format(args['model_name'], out_shape))
+        if out_config['data_type'] != 'TYPE_INT32':
+            raise pb_utils.TritonModelException(
+                """the model `{}` requires the data_type of 'OUT' to be
+                'TYPE_INT32', got {}""".format(args['model_name'],
+                                               out_config['data_type']))
+
+        self.inflight_thread_count = 0
+        self.inflight_thread_count_lck = threading.Lock()
+
+    def execute(self, requests):
+        """`execute` MUST be implemented in every Python model. `execute`
+        function receives a list of pb_utils.InferenceRequest as the only
+        argument. This function is called when an inference request is made
+        for this model. The request.get_response_sender() must be used to
+        get an InferenceResponseSender object associated with the request.
+        Use the InferenceResponseSender.send() and InferenceResponseSender.end()
+        calls to send responses and indicate no responses will be sent for
+        the corresponding request respectively. If there is an error, you can
+        set the error argument when creating a pb_utils.InferenceResponse
+
+        Parameters
+        ----------
+        requests : list
+          A list of pb_utils.InferenceRequest
+
+        Returns
+        -------
+        None
+        """
+
+        for request in requests:
+            self.process_request(request)
+
+        return None
```

Comment:
Why `repeat` and `square` models don't wait for the execution? Doesn't that mean there can be multiple concurrent executions even with 1 instance?

Reply:
>  Doesn't that mean there can be multiple concurrent executions even with 1 instance?

That is true. The behavior is documented here: https://github.com/triton-inference-server/python_backend/blob/tanmayv-decoupled/examples/decoupled/repeat_model.py#L59

Once the issue thread exits it can re-enter if there are new requests in the model. There can be multiple responses from multiple requests in-flight from the model/server. I can make the documentation more explicit. 

See here for how we do it in our C++ example: 
repeat: https://github.com/triton-inference-server/repeat_backend/blob/main/src/repeat.cc#L488-L500
square: https://github.com/triton-inference-server/square_backend/blob/main/src/square.cc#L360-L372

Reply:
I see, it makes sense to showcase the flexibility of the decouple style: execution / request / response are not closely coupled, and their lifecycle can be managed in arbitrary order.

Should probably point out that this bypasses the instance limitation and shouldn't be adopted to actual backends as it can result in over-committing work in the backend

Reply:
Added a comment in the execute function to explain the pitfall.